src/main.c

```diff
@@ -730,19 +733,41 @@ static int stop_components(void) {
 
   INFO(MSG_STOPING_COMPS);
 
-  int rc = 0;
-
   for (size_t i = 0; i < zl_context.child_count; i++) {
-    if (stop_component(&zl_context.children[i])) {
-      rc = -1;
+      pid_t pgid = -zl_context.children[i].pid;
+      zl_comp_t *comp = &zl_context.children[i];
+      comp->clean_stop = true;
+      kill(pgid, SIGTERM);
+  }
+
+  int wait_time = 0;
+  int max_wait_time = 20 * 1000;
+  bool all_exit = false;
+  while (!all_exit  && wait_time < max_wait_time) {
+    all_exit = true;
+    for (size_t i = 0; i < zl_context.child_count; i++) {
+      if (zl_context.children[i].pid > 0) {
+        all_exit = false;
+      }
     }
+    usleep(300000);
+    wait_time += 300;
   }
-
-  if (rc) {
-    WARN(MSG_NOT_ALL_STOPPED);
-  } else {
-    INFO(MSG_COMPS_STOPPED);
-  }
+  
+  for (size_t i = 0; i < zl_context.child_count; i++) {
+    if (zl_context.children[i].pid > 0) {
```

Comment:
There are many usage of `zl_context.children[i]` below, define it as a variable.